File: frontend/src/lib/rbac.js
export const INTERNAL_ROLES = {
  SUPER_ADMIN: 'SUPER_ADMIN',
  ADMIN: 'ADMIN',
  OPS: 'OPS',
  SUPPORT: 'SUPPORT',
  UNKNOWN: 'UNKNOWN',
};

export const PERMISSIONS = {
  SUPPORT_VIEW: 'SUPPORT_VIEW',
  OPS_ACTIONS: 'OPS_ACTIONS',
  CREDIT_ADJUSTMENT: 'CREDIT_ADJUSTMENT',
};

const SET_SUPPORT_VIEW = new Set([
  INTERNAL_ROLES.SUPPORT,
  INTERNAL_ROLES.OPS,
  INTERNAL_ROLES.ADMIN,
  INTERNAL_ROLES.SUPER_ADMIN,
]);

const SET_OPS_ACTIONS = new Set([
  INTERNAL_ROLES.OPS,
  INTERNAL_ROLES.ADMIN,
  INTERNAL_ROLES.SUPER_ADMIN,
]);

const SET_CREDIT_ADJUSTMENT = new Set([
  INTERNAL_ROLES.ADMIN,
  INTERNAL_ROLES.SUPER_ADMIN,
]);

const PERMISSION_MATRIX = {
  [PERMISSIONS.SUPPORT_VIEW]: SET_SUPPORT_VIEW,
  [PERMISSIONS.OPS_ACTIONS]: SET_OPS_ACTIONS,
  [PERMISSIONS.CREDIT_ADJUSTMENT]: SET_CREDIT_ADJUSTMENT,
};

export function normalizeRole(rawRole) {
  const r = (rawRole || '').trim();
  if (!r) return INTERNAL_ROLES.UNKNOWN;

  const key = r.replace(/[-_]/g, ' ').trim().toUpperCase();

  if (key === 'SUPER ADMIN' || key === 'SUPERADMIN') return INTERNAL_ROLES.SUPER_ADMIN;

  // Admin aliases (treat Tenant Admin as ADMIN)
  if (
    key === 'TENANT ADMIN' ||
    key === 'TENANTADMIN' ||
    key === 'TENANT ADMINISTRATOR' ||
    key === 'ADMIN' ||
    key === 'ADMINISTRATOR'
  ) {
    return INTERNAL_ROLES.ADMIN;
  }

  // Ops aliases
  if (key === 'OPS' || key === 'OPERATIONS' || key === 'OPERATION' || key === 'OP') return INTERNAL_ROLES.OPS;

  // Support aliases
  if (key === 'SUPPORT' || key === 'CS' || key === 'CUSTOMER SUPPORT' || key === 'CUSTOMERSUPPORT') {
    return INTERNAL_ROLES.SUPPORT;
  }

  return INTERNAL_ROLES.UNKNOWN;
}

export function getEffectiveRole(admin) {
  if (admin?.is_platform_owner) return INTERNAL_ROLES.SUPER_ADMIN;
  return normalizeRole(admin?.role);
}

export function hasPermission(adminOrRole, permission) {
  const role =
    typeof adminOrRole === 'string'
      ? normalizeRole(adminOrRole)
      : getEffectiveRole(adminOrRole);

  const set = PERMISSION_MATRIX[permission];
  return !!set && set.has(role);
}

export function getAdminFromStorage() {
  try {
    const raw = localStorage.getItem('admin_user');
    if (!raw) return null;
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

File: temp_extract/backend/app/services/rbac.py
from __future__ import annotations

from fastapi import HTTPException

from app.models.sql_models import AdminUser


# Canonical roles for Player Ops RBAC
# NOTE: We normalize raw role strings to avoid "string drift" causing accidental 403s.
ROLE_SUPPORT = "Support"
ROLE_OPS = "Ops"
ROLE_ADMIN = "Admin"
ROLE_SUPER_ADMIN = "Super Admin"


def normalize_role(raw_role: str | None) -> str:
    r = (raw_role or "").strip()
    if not r:
        return ""

    key = r.replace("-", " ").replace("_", " ").strip().upper()

    # Admin aliases (treat Tenant Admin as ADMIN for P0 "least surprise")
    if key in {"TENANT ADMIN", "TENANTADMIN", "TENANT ADMINISTRATOR", "ADMIN", "ADMINISTRATOR"}:
        return ROLE_ADMIN

    # Ops aliases
    if key in {"OPS", "OPERATIONS", "OPERATION", "OP"}:
        return ROLE_OPS

    # Support aliases
    if key in {"SUPPORT", "CS", "CUSTOMER SUPPORT", "CUSTOMERSUPPORT"}:
        return ROLE_SUPPORT

    # Super admin aliases
    if key in {"SUPER ADMIN", "SUPERADMIN"}:
        return ROLE_SUPER_ADMIN

    # If unknown, keep original trimmed string
    return r


def canonical_role(admin: AdminUser) -> str:
    """Return the canonical role label used by RBAC."""

    return normalize_role(getattr(admin, "role", None))


def require_role_any(admin: AdminUser, allowed: set[str]) -> None:
    role = canonical_role(admin)

    # Super Admin / Owner override
    if role == ROLE_SUPER_ADMIN or getattr(admin, "is_platform_owner", False):
        return

    if role not in allowed:
        raise HTTPException(status_code=403, detail={"error_code": "FORBIDDEN"})


# Policy helpers

def require_support_view(admin: AdminUser) -> None:
    # Everyone with a role can view player audit/notes.
    require_role_any(admin, {ROLE_SUPPORT, ROLE_OPS, ROLE_ADMIN, ROLE_SUPER_ADMIN})


def require_ops(admin: AdminUser) -> None:
    require_role_any(admin, {ROLE_OPS, ROLE_ADMIN, ROLE_SUPER_ADMIN})


def require_admin(admin: AdminUser) -> None:
    require_role_any(admin, {ROLE_ADMIN, ROLE_SUPER_ADMIN})
